hoomd/logging.py

```diff
@@ -1,74 +1,139 @@
+            if isinstance(flag, str):
+                return cls[flag]
             else:
-                return func
-        if func is None:
-            return helper
-        else:
-            return helper(func)
-
-    def __new__(cls, name, base, dct):
-        new_cls = super().__new__(cls, name, base, dct)
-        log_dict = dict()
-        for name, flag in cls._meta_export_dict.items():
-            log_dict[name] = LoggerQuantity(name, new_cls, flag)
-        if hasattr(new_cls, '_export_dict'):
-            old_dict = deepcopy(new_cls._export_dict)
-            for key, value in old_dict.items():
-                old_dict[key] = value.update_cls(new_cls)
-            old_dict.update(log_dict)
-            new_cls._export_dict = old_dict
-        else:
-            new_cls._export_dict = log_dict
-        cls._meta_export_dict = dict()
-        return new_cls
+                return flag
 
+        return reduce(
+            lambda x, y: from_str(x) | from_str(y), flags, TypeFlags.NONE)
 
-def generate_namespace(cls):
-    ns = tuple(cls.__module__.split('.') + [cls.__name__])
-    if ns[0] == 'hoomd':
-        return ns[1:]
-    else:
-        return ns
+    @classmethod
+    def _get_string_list(cls, flag):
+        return [mem.name for mem in cls.__members__.values() if mem in flag]
+
+
+TypeFlags.ALL = TypeFlags.any(TypeFlags.__members__.values())
 
 
-class LoggerQuantity:
+# function defined here to ensure that each class of type Loggable will have a
+# loggables property
+def _loggables(self):
+    """dict[str, str]: Return a name, flag mapping of loggable quantities."""
+    return {name: quantity.flag.name
+            for name, quantity in self._export_dict.items()}
+
+
+class _LoggableEntry:
+    """Stores entries for _Loggable's store of a class's loggable quantities."""
+    def __init__(self, flag, default):
+        self.flag = flag
+        self.default = default
+
+
+class _LoggerQuantity:
     """The information to automatically log to a `hoomd.logging.Logger`.
 
     Args:
         name (str): The name of the quantity.
         cls (``class object``): The class that the quantity comes from.
-        flag (str): The type of quantity it is. Valid values include
-            scalar, multi (array-like), particle (per particle quantity), bond,
-            angle, dihedral, constraint, pair, dict (a mapping of multiple
-            logged quantity names with their values), and object.
+        flag (str or TypeFlags, optional): The type of quantity it is.
+            Valid values are given in the `hoomd.logging.TypeFlags`
+            documentation.
 
     Note:
         For users, this class is meant to be used in conjunction with
         `hoomd.custom.Action` for exposing loggable quantities for custom user
         actions.
     """
-    def __init__(self, name, cls, flag='scalar'):
-        self.name = name
+
+    def __init__(self, name, cls, flag='scalar', default=True):
+        self.name = str(name)
```

Comment:
Is this cast necessary because name could be an int? Otherwise it seems cleaner to just assume it's a string and error somewhere if it's not.

Reply:
That is fine to remove.